### PR TITLE
cnf network: add qinq second PR multi-cvlans

### PR DESCRIPTION
**Pull request by gkopels**
Adding 4 test cases.
2 multi-cvlans
1 negative test
1 simultaneous 802.1ad and 802.1q

### CODE REVIEW

File: tests/cnf/core/network/sriov/tests/qinq.go [hunk 1]
```diff
@@ -185,28 +212,129 @@ var _ = Describe("QinQ", Ordered, Label(tsparams.LabelQinQTestCases), ContinueOn
...
+
+				By("Define and create a 802.1AD  client container")
+				annotation = defineNetworkAnnotation(srIovNetworkDot1AD, nadCVLAN100, false)
+				clientDotADPod := createClientTestPod(clientNameDot1ad, workerNodeList[0].Definition.Name, annotation)
+
+				By("Define and create a 802.1Q server container")
+				annotation = defineNetworkAnnotation(srIovNetworkDot1Q, nadCVLAN101, true)
+				serverDotQPod := createServerTestPod(serverNameDot1q, workerNodeList[0].Definition.Name, testCmdNet2,
+					annotation)
+				By("Define and create a 802.1AD client container")
```
Comment (mlguerrero12) [inline tests/cnf/core/network/sriov/tests/qinq.go:312]: this should be .1Q, right?

File: tests/cnf/core/network/sriov/tests/qinq.go [hunk 2]
```diff
@@ -254,28 +379,65 @@ var _ = Describe("QinQ", Ordered, Label(tsparams.LabelQinQTestCases), ContinueOn
...
-				By("Validate IPv6 tcp traffic and dot1ad encapsulation from the client to server.")
-				validateTCPTraffic(clientPod, serverIPV6IP.String())
+				By("Validate IPv4 and IPv6 tcp traffic and dot1q encapsulation from the client to server")
+				validateTCPTraffic(clientPod, intNet2, serverIPAddressesNet2)
 
 				By("Validate that the TCP traffic is double tagged")
 				readAndValidateTCPDump(tcpDumpContainer, tcpDumpReadFileCMD, tcpDumpDot1QOutput)
 			})
+
+		It("Verify network traffic over a double tagging QinQ tunnel with multiple C-VLANs using the same S-VLAN",
```
Comment (mlguerrero12) [inline tests/cnf/core/network/sriov/tests/qinq.go:402]: double tagging QinQ is too general - it's better 802.1q/802.1q or two times 802.1q

File: tests/cnf/core/network/sriov/tests/qinq.go (hunk 1 above)
Comment (gkopels) [reply tests/cnf/core/network/sriov/tests/qinq.go:312]: Correct fixed

File: tests/cnf/core/network/sriov/tests/qinq.go (hunk 2 above)
Comment (gkopels) [reply tests/cnf/core/network/sriov/tests/qinq.go:402]: How about double tagged 802.1Q ?

File: tests/cnf/core/network/sriov/tests/qinq.go [hunk 3]
```diff
@@ -44,24 +44,45 @@ var _ = Describe("QinQ", Ordered, Label(tsparams.LabelQinQTestCases), ContinueOn
...
 		promiscVFCommand            string
 		srIovInterfacesUnderTest    []string
 		sriovDeviceID               string
 		switchCredentials           *sriovenv.SwitchCredentials
 		switchConfig                *netconfig.NetworkConfig
 		switchInterfaces            []string
+		serverIPAddressesNet2       []string
+		serverIPAddressesNet3       []string
+		clientIPAddressesNet2       []string
+		clientIPAddressesNet3       []string
```
Comment (evgenLevin) [inline tests/cnf/core/network/sriov/tests/qinq.go:74]: replace by
```
	serverIPAddressesNet2 = []string{serverIPV4IP.String(), serverIPV6IP.String()}
	serverIPAddressesNet3 = []string{serverIPV4IP2.String(), serverIPV6IP2.String()}
	clientIPAddressesNet2 = []string{tsparams.ClientIPv4IPAddress, tsparams.ClientIPv6IPAddress}
	clientIPAddressesNet3 = []string{tsparams.ClientIPv6IPAddress2, tsparams.ClientIPv6IPAddress2}
```

**Review by evgenLevin — APPROVED** [review]

**Comment by ajaggapa** [comment]
lgtm

**Review by ajaggapa — APPROVED** [review]

File: tests/cnf/core/network/sriov/tests/qinq.go (hunk 3 above)
Comment (gkopels) [reply tests/cnf/core/network/sriov/tests/qinq.go]: changed

**Review by ajaggapa — APPROVED** [review]

File: tests/cnf/core/network/sriov/tests/qinq.go [hunk 4]
```diff
@@ -348,40 +505,57 @@ func createPromiscuousClient(nodeName string, tcpDumpCMD []string) *pod.Builder
...
-		NetConfig.CnfNetTestContainer).DefineOnNode(nodeName).WithSecondaryNetwork(annotation).
+		NetConfig.CnfNetTestContainer).DefineOnNode(nodeName).WithSecondaryNetwork(networkAnnotation).
 		WithPrivilegedFlag().CreateAndWaitUntilRunning(netparam.DefaultTimeout)
 	Expect(err).ToNot(HaveOccurred(), "Failed to define and run default client")
 
 	return clientBuild
 }
 
-func defineNetworkAnnotation(sVlan, cVlan string, ipAddress []string) []*multus.NetworkSelectionElement {
+func defineNetworkAnnotation(sVlan, cVlan string, server bool, cVlan2 ...string) []*multus.NetworkSelectionElement {
```
Comment (clobrano) [inline tests/cnf/core/network/sriov/tests/qinq.go:530]: The name `cVlan2` could be more descriptive. I am not familiar with the code and can't understand what this variable is for.

File: tests/cnf/core/network/sriov/tests/qinq.go [hunk 5]
```diff
@@ -348,40 +505,57 @@ func createPromiscuousClient(nodeName string, tcpDumpCMD []string) *pod.Builder
...
-	annotation := defineNetworkAnnotation(sVlan, cVlan, ipAddress)
-
-	serverCmd := []string{"bash", "-c", "sleep 5; testcmd -interface net2 -protocol tcp -port 4444 -listen"}
 	serverBuild, err := pod.NewBuilder(APIClient, name, tsparams.TestNamespaceName,
-		NetConfig.CnfNetTestContainer).DefineOnNode(nodeName).WithSecondaryNetwork(annotation).
-		RedefineDefaultCMD(serverCmd).WithPrivilegedFlag().CreateAndWaitUntilRunning(netparam.DefaultTimeout)
-	Expect(err).ToNot(HaveOccurred(), "Failed to define and run server container")
+		NetConfig.CnfNetTestContainer).DefineOnNode(nodeName).WithSecondaryNetwork(networkAnnotation).
+		RedefineDefaultCMD(command).WithPrivilegedFlag().CreateAndWaitUntilRunning(netparam.DefaultTimeout)
+	Expect(err).ToNot(HaveOccurred(), "Failed to define and run default client")
```
Comment (clobrano) [inline tests/cnf/core/network/sriov/tests/qinq.go:514]: This rings a bell to me. The function's name suggests this is to "create a server", while the error message is about a failure creating "a client". Is that right?

**Review by clobrano — CHANGES_REQUESTED** [review]
Sorry. My review system got the wrong PR and I didn't realize until it was too late. Ignore this.

**Pull request merged**